File: Medium/Subarrays with sum K/subarrays-with-sum-k.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;

class GFG{
    public static void main(String args[])throws IOException
    {
        BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine());
        while(t-- > 0)
        {
            int N = Integer.parseInt(read.readLine());
            String input_line[] = read.readLine().trim().split("\\s+");
            int Arr[]= new int[N];
            for(int i = 0; i < N; i++)
                Arr[i] = Integer.parseInt(input_line[i]);
            int k = Integer.parseInt(read.readLine());
            
            Solution ob = new Solution();
            System.out.println(ob.findSubArraySum(Arr, N, k));
        }
    }
}
// } Driver Code Ends


//User function Template for Java

class Solution
{
    static int findSubArraySum(int Arr[], int N, int k)
    {
        // code here
        long[] prefix = new long[N+1];
        int count = 0;

        HashMap<Long, Integer> m = new HashMap<>();
        m.put(0L, 1); // Initial frequency of sum 0

        for (int i = 0; i < N; i++) {
            prefix[i+1] = prefix[i] + Arr[i];
            m.put(prefix[i+1], m.getOrDefault(prefix[i+1], 0) + 1);
        }

        for (int i = 0; i < N; i++) {
            m.put(prefix[i], m.get(prefix[i]) - 1);
            count += m.getOrDefault(prefix[i] + k, 0);
        }

        return count;
    }
}

File: Difficulty: Medium/Subarrays with sum K/subarrays-with-sum-k.cpp
//{ Driver Code Starts
// Initial Template for C++
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function Template for C++

class Solution {
  public:
    int countSubarrays(vector<int> &arr, int k) {
        // code here
        int n=arr.size();
        vector<int> pre(n+1,0);
        unordered_map<int,int> m;
        m[0]++;
        int ans=0;
        for(int i=0;i<n;i++){
            pre[i+1]=(pre[i]+arr[i]);
            ans+=(m[pre[i+1]-k]);
            m[pre[i+1]]++;
        }
        return ans;
    }   
};

//{ Driver Code Starts.

int main() {

    int t;
    cin >> t;
    cin.ignore();
    while (t--) {

        string ks;
        getline(cin, ks);
        int k = stoi(ks);
        vector<int> arr;
        string input;
        getline(cin, input);
        stringstream ss(input);
        int number;
        while (ss >> number) {
            arr.push_back(number);
        }
        Solution obj;
        cout << obj.countSubarrays(arr, k);
        cout << endl;
        cout << "~"
             << "\n";
    }

    return 0;
}

// } Driver Code Ends